lib/googleauth/external_account/aws_credentials.rb

```diff
@@ -0,0 +1,346 @@
+# limitations under the License.
+
+require "time"
+require "googleauth/external_account/base_credentials"
+
+module Google
+  # Module Auth provides classes that provide Google-specific authorization
+  # used to access Google APIs.
+  module Auth
+    # Authenticates requests using External Account credentials, such
+    # as those provided by the AWS provider.
+    module ExternalAccount
+      # This module handles the retrieval of credentials from Google Cloud
+      # by utilizing the AWS EC2 metadata service and then exchanging the
+      # credentials for a short-lived Google Cloud access token.
+      class AwsCredentials
+        include Google::Auth::ExternalAccount::BaseCredentials
+        extend CredentialsLoader
+
+        # Will always be None, but method still gets used.
+        attr_reader :client_id
+
+        def initialize options = {}
+          base_setup options
+
+          @audience = options[:audience]
+          @credential_source = options[:credential_source] || {}
+          @environment_id = @credential_source["environment_id"]
+          @region_url = validate_metadata_server @credential_source["region_url"], "region_url"
+          @credential_verification_url = validate_metadata_server @credential_source["url"], "url"
+          @regional_cred_verification_url = @credential_source["regional_cred_verification_url"]
+          @imdsv2_session_token_url = validate_metadata_server @credential_source["imdsv2_session_token_url"],
+                                                               "imdsv2_session_token_url"
+
+          # These will be lazily loaded when needed, or will raise an error if not provided
+          @region = nil
+          @request_signer = nil
+        end
+
+        private
+
+        def validate_metadata_server url, name
+          return nil if url.nil?
+          host = URI(url).host
+          raise "Invalid host #{host} for #{name}." unless ["169.254.169.254", "[fd00:ec2::254]"].include? host
+          url
+        end
+
+        def get_aws_resource url, name, data: nil, headers: {}
+          begin
+            unless [nil, url].include? @imdsv2_session_token_url
+              headers["x-aws-ec2-metadata-token"] = get_aws_resource(
+                @imdsv2_session_token_url,
+                "Session Token",
+                headers: { "x-aws-ec2-metadata-token-ttl-seconds": "300" }
+              ).body
+            end
+
+            response = if data
+                         headers["Content-Type"] = "application/json"
+                         connection.post url, data, headers
+                       else
+                         connection.get url, nil, headers
+                       end
+
+            raise Faraday::Error unless response.success?
+            response
+          rescue Faraday::Error
+            raise "Failed to retrieve AWS #{name}."
+          end
+        end
+
+        def retrieve_subject_token!
```

Comment:
Can we please document what these methods do? May be like we have in python library too?